vmicore/src/vmi/InterruptEventSupervisor.cpp

```diff
@@ -0,0 +1,268 @@
+#include "VmiException.h"
+#include <memory>
+
+namespace VmiCore
+{
+    namespace
+    {
+        InterruptEventSupervisor* interruptEventSupervisor = nullptr;
+        const std::string loggerName = std::filesystem::path(__FILE__).filename().stem();
+    }
+
+    InterruptEventSupervisor::InterruptEventSupervisor(std::shared_ptr<ILibvmiInterface> vmiInterface,
+                                                       std::shared_ptr<ISingleStepSupervisor> singleStepSupervisor,
+                                                       std::shared_ptr<ILogging> loggingLib)
+        : vmiInterface(std::move(vmiInterface)),
+          singleStepSupervisor(std::move(singleStepSupervisor)),
+          loggingLib(std::move(loggingLib)),
+          logger(NEW_LOGGER(this->loggingLib))
+    {
+        interruptEventSupervisor = this;
+    }
+
+    InterruptEventSupervisor::~InterruptEventSupervisor() noexcept
+    {
+        interruptEventSupervisor = nullptr;
+    }
+
+    void InterruptEventSupervisor::initialize()
+    {
+        SETUP_INTERRUPT_EVENT(&event, _defaultInterruptCallback);
+        event.interrupt_event.reinject = DONT_REINJECT_INTERRUPT;
+        event.interrupt_event.insn_length = 1;
+        vmiInterface->registerEvent(event);
+        singleStepSupervisor->initializeSingleStepEvents();
+        singleStepCallbackFunction = SingleStepSupervisor::createSingleStepCallback(
+            shared_from_this(), &InterruptEventSupervisor::singleStepCallback);
+    }
+
+    void InterruptEventSupervisor::teardown()
+    {
+        clearInterruptEventHandling();
+        singleStepSupervisor->teardown();
+    }
+
+    std::shared_ptr<IBreakpoint> InterruptEventSupervisor::createBreakpoint(
+        uint64_t targetVA, uint64_t processCr3, const std::function<BpResponse(IInterruptEvent&)>& callbackFunction)
+    {
+
+        auto targetPA = vmiInterface->convertVAToPA(targetVA, processCr3);
+        auto breakpoint = std::make_shared<Breakpoint>(
+            targetPA,
+            [supervisor = weak_from_this()](Breakpoint* bp)
+            {
+                if (auto supervisorShared = supervisor.lock())
+                {
+                    supervisorShared->deleteBreakpoint(bp);
+                }
+            },
+            callbackFunction);
+
+        std::scoped_lock guard(lock);
+        // Register new INT3
+        if (!breakpointsByPA.contains(targetPA))
+        {
+            vmiInterface->flushV2PCache(LibvmiInterface::flushAllPTs);
+            vmiInterface->flushPageCache();
+            storeOriginalValue(targetPA);
+            enableEvent(targetPA);
+
+            auto targetGFN = targetPA >> PagingDefinitions::numberOfPageIndexBits;
+            // One PageGuard guards a whole memory page on which several interrupts may reside.
+            if (!interruptGuardsByGFN.contains(targetGFN))
+            {
+                interruptGuardsByGFN[targetGFN] = createPageGuard(targetVA, processCr3, targetGFN);
```

Comment:
only for Windows, right?

Reply:
Userspace breakpoints should be hidden on Linux as well, should they not?